config/config_partial.go

```diff
@@ -0,0 +1,190 @@
+)
+
+// terragruntInclude is a struct that can be used to only decode the include block.
+type terragruntInclude struct {
+	Include *IncludeConfig `hcl:"include,block"`
+	Remain  hcl.Body       `hcl:",remain"`
+}
+
+// terragruntDependencies is a struct that can be used to only decode the dependencies block.
+type terragruntDependencies struct {
+	Dependencies *ModuleDependencies `hcl:"dependencies,block"`
+	Remain       hcl.Body            `hcl:",remain"`
+}
+
+// terragruntTerraform is a struct that can be used to only decode the terraform block
+type terragruntTerraform struct {
+	Terraform *TerraformConfig `hcl:"terraform,block"`
+	Remain    hcl.Body         `hcl:",remain"`
+}
+
+// terragruntFlags is a struct that can be used to only decode the flag attributes (skip and prevent_destroy)
+type terragruntFlags struct {
+	PreventDestroy *bool    `hcl:"prevent_destroy,attr"`
+	Skip           *bool    `hcl:"skip,attr"`
+	Remain         hcl.Body `hcl:",remain"`
+}
+
+func PartialParseConfigFile(
+	filename string,
+	terragruntOptions *options.TerragruntOptions,
+	include *IncludeConfig,
+	decodeList []string,
+) (*TerragruntConfig, error) {
+	configString, err := util.ReadFileAsString(filename)
+	if err != nil {
+		return nil, err
+	}
+
+	config, err := PartialParseConfigString(configString, terragruntOptions, include, filename, decodeList)
+	if err != nil {
+		return nil, err
+	}
+
+	return config, nil
+}
+
+// ParitalParseConfigString partially parses and decodes the provided string. Which blocks/attributes to decode is
+// controlled by the function parameters.
+// Note that the following blocks are always decoded:
+// - locals
+// - include
+// Note also that the following blocks are never decoded in a partial parse:
+// - inputs
+func PartialParseConfigString(
+	configString string,
+	terragruntOptions *options.TerragruntOptions,
+	includeFromChild *IncludeConfig,
+	filename string,
+	decodeList []string,
+) (*TerragruntConfig, error) {
+	// Parse the HCL string into an AST body that can be decoded multiple times later without having to re-parse
+	parser := hclparse.NewParser()
+	file, err := parseHcl(parser, configString, filename)
+	if err != nil {
+		return nil, err
+	}
+
+	// Evaluate all the expressions in the locals block separately and generate the variables list to use in the
+	// evaluation context.
+	locals, err := evaluateLocalsBlock(terragruntOptions, parser, file, filename)
+	if err != nil {
+		return nil, err
+	}
+
+	// Decode just the `include` block, and verify that it's allowed here
+	terragruntInclude, err := decodeAsTerragruntInclude(file, filename, terragruntOptions, locals)
+	if err != nil {
+		return nil, err
+	}
+	includeForDecode, err := getIncludedConfigForDecode(terragruntInclude, terragruntOptions, includeFromChild)
+	if err != nil {
+		return nil, err
+	}
+
+	output := TerragruntConfig{IsPartial: true}
+
+	for _, decode := range decodeList {
+		switch decode {
+		case "dependencies":
```

Comment:
A bit confused on what we're doing in this loop?

Reply:
This makes more sense if you understand the original root problem implied in https://github.com/gruntwork-io/terragrunt/pull/828#discussion_r312653873

Reply:
See https://github.com/gruntwork-io/terragrunt/pull/828#discussion_r312662399. Now imagine the following modified config for `sql`:

```
locals = {
  vpc_path = "../vpc"
}

dependencies {
  paths = [local.vpc_path]
}

inputs = {
  vpc_id = get_output(local.vpc_path, "vpc_id")
}
```

In the original implementation of the parsing logic for `apply-all`, we will still fail because we will naively parse the entire config file, including the `inputs`, regardless of whether or not we needed it in the dependency tree builder (creating the `Stack` struct).

Looking at the routine for building up the `Stack` struct, the only sections of the config that we need at that moment is the `terraform` and `dependencies` block. So this function and for loop is providing a routine that allows you to partially parse the terragrunt config so that you only read in those blocks, and ignore the rest.

`PartialParseConfigString(..., []string{"dependencies", "terraform"})` only parses the two blocks `dependencies` and `terraform` of the config (plus `include` and `locals`, which can modify those), and ignores the rest. In this way, we can extract the information we need without parsing the entire config, which may not be ready for parsing. Later on, when we actually run the `terraform apply`, the config is parsed again in full.

The forloop is just an implementation detail of the partial parsing, so that you can better control which sections are decoded. I couldn't find a clean way to programmatically build up a struct in a clean way. So instead, we conditionally pass in different structs that only extract the pieces we need from the config.

Reply:
Should those be enums rather than strings?

Reply:
Switched to using enums.